File: game-engine/src/main/java/org/luxons/sevenwonders/game/moves/CardFromHandMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;

public abstract class CardFromHandMove extends Move {

    CardFromHandMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public void validate(Table table, List<Card> playerHand) throws InvalidMoveException {
        if (!playerHand.contains(getCard())) {
            throw new InvalidMoveException(
                    String.format("Player %d does not have the card '%s' in his hand", getPlayerIndex(),
                            getCard().getName()));
        }
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/CardFromHandMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.api.Table
import org.luxons.sevenwonders.game.cards.Card

abstract class CardFromHandMove internal constructor(playerIndex: Int, card: Card, move: PlayerMove) :
    Move(playerIndex, card, move) {

    @Throws(InvalidMoveException::class)
    override fun validate(table: Table, playerHand: List<Card>) {
        if (!playerHand.contains(card)) {
            throw InvalidMoveException("Player $playerIndex does not have the card '${card.name}' in his hand")
        }
    }
}
